### ISSUE TITLE
Add a way to check if the WebSocket connection between the frontend and backend via the WebUI bridge has been successfully established or not

### ISSUE THREAD

**Issue by Arc8ne** [Juff-Ma/WebUI.NET#3]
While trying out this library in a new project, I realized that there is no way to check or be notified when the WebSocket connection between the frontend and backend (which is required in order to be able to successfully call backend functions from the frontend) via the WebUI bridge has been successfully established.

Could such a way be added to this library (i.e. in the form of an event or public field exposed by the `webui` object that is accessible in frontend JS scripts)?

**Comment by Juff-Ma** [comment]
WebUI.NET only contains bindings for webui itself, in particular version 2.4. The main c project (http://webui.me) is currently in the beta phase of version 2.5, you could request the feature there and it would then be implemented in the next WebUI.NET version.

**Comment by Juff-Ma** [comment]
Also I'm not that familiar with the bridge itself, maybe there is such functionality built in and I don't know how to access it. However the app should not be accessible if the bridge is not yet connected, the bridge pauses all inputs if it's disconnected, at least as far as I know.

**Comment by Arc8ne** [comment]
Sorry but I am confused, after taking a look at the `webui.ts` file used by the original WebUI project, there already seems to be ways to do what I asked for initially via the `isConnected` and `setEventCallback` methods of the `WebuiBridge` class defined in the said TS file, however, the 'virtual' `webui.js` script injected/provided by this library does not have those 2 methods defined in its `WebuiBridge` class.

Why is that the case?

**Comment by Arc8ne** [comment]
> Also I'm not that familiar with the bridge itself, maybe there is such functionality built in and I don't know how to access it. However the app should not be accessible if the bridge is not yet connected, the bridge pauses all inputs if it's disconnected, at least as far as I know.

In that case, how will the frontend know when the bridge is connected so that it can call a backend function without errors?

**Comment by Juff-Ma** [comment]
You probably looked at the code of version 2.5.0-beta2, the version WebUI.NET currently provides bindings for is the last non-beta version (e.g. 2.4.2). The code of the 2.4.2 bridge (webui_bridge.ts) does not seem to actively expose this functionality as a function.

**Comment by Juff-Ma** [comment]
> > Also I'm not that familiar with the bridge itself, maybe there is such functionality built in and I don't know how to access it. However the app should not be accessible if the bridge is not yet connected, the bridge pauses all inputs if it's disconnected, at least as far as I know.
> 
> In that case, how will the frontend know when the bridge is connected so that it can call a backend function without errors?

If I remember correctly, you could check if the function you want to call is already defined, because the bridge will define them once it discovers them, which requires the bridge to be connected. So, for example, if your callback on the backend is called "get_id_with_name" you could do a check like `if (webui.get_id_with_name) webui.get_id_with_name("name1")`

**Comment by Arc8ne** [comment]
> > > Also I'm not that familiar with the bridge itself, maybe there is such functionality built in and I don't know how to access it. However the app should not be accessible if the bridge is not yet connected, the bridge pauses all inputs if it's disconnected, at least as far as I know.
> > 
> > 
> > In that case, how will the frontend know when the bridge is connected so that it can call a backend function without errors?
> 
> If I remember correctly, you could check if the function you want to call is already defined, because the bridge will define them once it discovers them, which requires the bridge to be connected. So, for example, if your callback on the backend is called "get_id_with_name" you could do a check like `if (webui.get_id_with_name) webui.get_id_with_name("name1")`

I see, so lets say that I only wanted the rest of my frontend code to run once the WebUI bridge has been connected, I could just execute a loop that waits until the backend function is defined on the `webui` object?

So for instance instead of this code:
```ts
function waitTillWebUiBridgeIsConnected(): Promise<void>
{
    return new Promise(
        (resolve) =>
        {
            document.addEventListener(
                "DOMContentLoaded",
                () =>
                {
                    webui.setEventCallback(
                        (e) =>
                        {
                            if (e == webui.event.CONNECTED)
                            {
                                resolve();
                            }
                        }
                    );
                }
            );
        }
    );
}
```

I would use the below code instead?
```ts
function waitTillWebUiBridgeIsConnected(): Promise<void>
{
    return new Promise(
        (resolve) =>
        {
            while (webui.nameOfBackendFunction == undefined)
            {
                // Do nothing.
            }

            resolve();
        }
    );
}
```

**Comment by Juff-Ma** [comment]
> > > > Also I'm not that familiar with the bridge itself, maybe there is such functionality built in and I don't know how to access it. However the app should not be accessible if the bridge is not yet connected, the bridge pauses all inputs if it's disconnected, at least as far as I know.
> > > 
> > > 
> > > In that case, how will the frontend know when the bridge is connected so that it can call a backend function without errors?
> > 
> > If I remember correctly, you could check if the function you want to call is already defined, because the bridge will define them once it discovers them, which requires the bridge to be connected. So, for example, if your callback on the backend is called "get_id_with_name" you could do a check like `if (webui.get_id_with_name) webui.get_id_with_name("name1")`
> 
> I see, so lets say that I only wanted the rest of my frontend code to run once the WebUI bridge has been connected, I could just execute a loop that waits until the backend function is defined on the `webui` object?

Yes you could do that, though make sure you include your script after webui.js as otherwise it could keep webui from initializing properly.

Edit: maybe also so a small sleep in the loop, doing those checks without sleeping in between could cause CPU ramp up.

**Comment by Arc8ne** [comment]
I see, thanks for the clarification. Will those new `WebuiBridge` methods (i.e. `setEventCallback`, `isConnected`) be added once version 2.5.0 of the original WebUI is no longer in beta (i.e. is stable)?

**Comment by Juff-Ma** [comment]
> I see, thanks for the clarification. Will those new `WebuiBridge` methods (i.e. `setEventCallback`, `isConnected`) be added once version 2.5.0 of the original WebUI is no longer in beta (i.e. is stable)?

Yes, I plan to update to version 2.5.0 when it's out of beta, WebUI.NET uses the same bridge as webui itself so the new methods would be added automatically.

**Comment by Arc8ne** [comment]
I see, thanks for letting me know about that. If you do not mind also clarifying, what is the difference between this library and the following library: https://github.com/salvadordf/WebUI4CSharp?

They both seem to be doing the same thing but I am still unsure about which library to proceed with for a project that I am working on.

**Comment by Juff-Ma** [comment]
> I see, thanks for letting me know about that. If you do not mind also clarifying, what is the difference between this library and the following library: https://github.com/salvadordf/WebUI4CSharp?
> 
> They both seem to be doing the same thing but I am still unsure about which library to proceed with for a project that I am working on.

The projects were both developed at the same time, we both didn't know the other one already existed. The main differences are the license (WebUI.NET being more strictly licensed using the MPL), that WebUI4CSharp already supports the 2.5 beta and that WebUI.NET is a more high level wrapper while WebUI4CSharp has more resemblance to the native c library.

**Comment by Juff-Ma** [comment]
Also WebUI.NET has precompiled SSL binaries, however those got some nasty problems in version 2.4 that are only fixed in version 2.5

**Comment by Arc8ne** [comment]
I see, thanks for the information.
